### PR TITLE
API: Fix cython exception handling for exported extern C functions

### LINKED ISSUE #19291: BUG,ENH: cython stubs should use `except (0|-1?)` etc. for integer returning functions
According to the docs `PyArray_IntpConverter` is supposed to just return `NPY_FAIL` (0) on failure and `NPY_SUCCESS` (1) otherwise. So to help manually raise an exception I write something like:
```cython
cdef np.PyArray_Dims shape
if not np.PyArray_IntpConverter(size, &shape):
	raise ValueError("`size` is not a valid argument.")
```

But this ends up raising a `SystemError` with the message: `<class 'ValueError'> returned a result with an error set`, Which ends up causing unittests to fail since the intended ValueError is not raised.
### Reproducing code example:
Passing a set for `size` instead of the expected tuple.

Is there a way to silence the `SystemError` and only raise the intended one?

### NumPy/Python version information:

```
1.19.0 3.6.12 (default, May 11 2021, 17:48:34) 
[GCC 10.2.0]
```

### PR DESCRIPTION
**Pull request by seberg**
This hopefully fixes them for all the functions currently in the `.pyd` files.  A surprising amount of them look like scary thing I wouldn't mind to just delete :).

Anyway, had looked at some Cython code today and remembered this.

Closes gh-19291

### CODE REVIEW

File: numpy/__init__.cython-30.pxd [hunk 1]
```diff
@@ -559,11 +559,11 @@ cdef extern from "numpy/arrayobject.h":
...
     #object PyArray_CastToType (ndarray, dtype, int)
-    int PyArray_CastTo (ndarray, ndarray)
-    int PyArray_CastAnyTo (ndarray, ndarray)
-    int PyArray_CanCastSafely (int, int)
-    npy_bool PyArray_CanCastTo (dtype, dtype)
-    int PyArray_ObjectType (object, int)
+    int PyArray_CastTo (ndarray, ndarray) except -1
+    int PyArray_CastAnyTo (ndarray, ndarray) except -1
+    int PyArray_CanCastSafely (int, int)  # writes errors
+    npy_bool PyArray_CanCastTo (dtype, dtype)  # writes errors
```
Comment (eric-wieser) [inline numpy/__init__.cython-30.pxd:565]: I think all these "writes errors" should either be `except *` (if you can't say anything about the return value), or `except? 0` (if errors always produce zero, but not all zeros are errors).

Comment (seberg) [reply numpy/__init__.cython-30.pxd:565]: Oh, "writes errors" here means it uses `PyErr_WriteUnraiseable`.  So I think for cython that should be as good as nothing.

Comment (eric-wieser) [reply numpy/__init__.cython-30.pxd:565]: Oh, I thought this meant "sets the exception but doesn't indicate it in the return value".

File: numpy/__init__.cython-30.pxd [hunk 2]
```diff
@@ -587,34 +587,34 @@ cdef extern from "numpy/arrayobject.h":
...
+    double PyArray_GetPriority (object, double)  # clears errors as of 1.25
     object PyArray_IterNew (object)
     object PyArray_MultiIterNew (int, ...)
 
-    int PyArray_PyIntAsInt (object)
+    int PyArray_PyIntAsInt (object) except? -1
     npy_intp PyArray_PyIntAsIntp (object)
-    int PyArray_Broadcast (broadcast)
+    int PyArray_Broadcast (broadcast) except -1
     void PyArray_FillObjectArray (ndarray, object)
```
Comment (eric-wieser) [inline numpy/__init__.cython-30.pxd:616]: Should this one (and probably a handful of other void functions) be `except *`? I assume this doesn't swallow errors.

Comment (seberg) [reply numpy/__init__.cython-30.pxd:616]: Arg, I didn't scan the `void` ones carefully, I guess.  This isn't designed to error, but clearly can...  (one of those bad functions).  (the xincref/decref family is also weird, I think they tend to clear, but probably not always... and they shouldn't be used anway, especially in cython)

**Comment by seberg** [comment]
As confirmed by Stefan Behnel in the issue, I think this should be rather safe to do.  Cython will insert error handling and any manual error handling simply becomes obsolete.

File: numpy/__init__.cython-30.pxd (hunk 1 above)
Comment (charris) [reply numpy/__init__.cython-30.pxd:565]: PyErr_WriteUnraiseable -> PyErr_WriteUnraisable

Just to help searching for the documentation :)

**Comment by charris** [comment]
Let's give this a shot. Thanks Sebastian.

**Comment by zoj613** [comment]
Thanks @seberg

**Comment by seberg** [comment]
@scoder is it still relevant to push such changes to cython (in which case I can make a PR)?

**Pull request merged**